File: MST/boj_1647.py
import heapq
import sys
input = sys.stdin.readline


def find(x):
    if x != parent[x]:
        parent[x] = find(parent[x])
    return parent[x]


def union(a, b):
    parentA = find(a)
    parentB = find(b)

    if parentA > parentB:
        parent[parentB] = parentA
    else:
        parent[parentA] = parentB

        
N, M = map(int, input().split())
parent = [i for i in range(N + 1)]
edges = []
for _ in range(M):
    u, v, w = map(int, input().split())
    heapq.heappush(edges, [w, u, v])

total_cost = 0
last_cost = 0

while edges:
    edge = heapq.heappop(edges)

    if find(edge[1]) != find(edge[2]):
        union(edge[1], edge[2])
        total_cost += edge[0]
        last_cost = edge[0]

print(total_cost - last_cost)

File: MST/boj_1647.java
package MST;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class boj_1647 {
    static int N, M;
    static int[] parent;

    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        PriorityQueue<Edge> edges = new PriorityQueue<>();
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());

            edges.offer(new Edge(u, v, w));
        }

        parent = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            parent[i] = i;
        }

        int totalCost = 0;
        int lastCost = 0;
        while (!edges.isEmpty()) {
            Edge edge = edges.poll();

            if (find(edge.start) != find(edge.end)) {
                union(edge.start, edge.end);
                totalCost += edge.cost;

                lastCost = edge.cost;
            }
        }

        System.out.println(totalCost - lastCost);
    }

    static int find(int x) {
        if (x != parent[x]) parent[x] = find(parent[x]);
        return parent[x];
    }

    static void union(int a, int b) {
        int parentA = find(a);
        int parentB = find(b);

        if (parentA > parentB) {
            parent[parentA] = parentB;
        } else {
            parent[parentB] = parentA;
        }
    }

    static class Edge implements Comparable<Edge> {
        int start;
        int end;
        int cost;

        public Edge(int start, int end, int cost) {
            this.start = start;
            this.end = end;
            this.cost = cost;
        }

        @Override
        public int compareTo(Edge o) {
            return Integer.compare(this.cost, o.cost);
        }
    }
}
